PubSub/Redis/RedisClient.php

```diff
@@ -0,0 +1,118 @@
+
+    public function __construct()
+    {
+        $this->apps     = collect(config('websockets.apps'));
+        $this->serverId = Str::uuid()->toString();
+    }
+
+    public function publish(string $appId, array $payload): bool
+    {
+        $payload['appId']    = $appId;
+        $payload['serverId'] = $this->serverId;
+        $this->publishClient->publish(self::REDIS_KEY, json_encode($payload));
+        return true;
+    }
+
+    public function subscribe(LoopInterface $loop): PubSubInterface
+    {
+        $this->loop = $loop;
+        [$this->publishClient, $this->subscribeClient] = Block\awaitAll([$this->publishConnection(), $this->subscribeConnection()], $this->loop);
+        return $this->publishClient;
+    }
+
+    protected function publishConnection(): PromiseInterface
+    {
+        $connectionUri = $this->getConnectionUri();
+        $factory       = new Factory($this->loop);
+        return $factory->createClient($connectionUri)->then(
+            function (Client $client) {
+                $this->publishClient = $client;
+                return $this;
+            }
+        );
+    }
+
+
+    protected function subscribeConnection(): PromiseInterface
+    {
+        $connectionUri = $this->getConnectionUri();
+        $factory       = new Factory($this->loop);
+        return $factory->createClient($connectionUri)->then(
+            function (Client $client) {
+                $this->subscribeClient = $client;
+                $this->onConnected();
+                return $this;
+            }
+        );
+    }
+
+    protected function getConnectionUri()
+    {
+        $name   = config('websockets.replication.connection') ?? 'default';
+        $config = config('database.redis.' . $name);
+        $host   = $config['host'];
+        $port   = $config['port'] ? (':' . $config['port']) : ':6379';
+
+        $query = [];
+        if ($config['password']) {
+            $query['password'] = $config['password'];
+        }
+        if ($config['database']) {
+            $query['database'] = $config['database'];
+        }
+        $query = http_build_query($query);
+
+        return "redis://$host$port" . ($query ? '?' . $query : '');
+    }
+
+    protected function onConnected()
+    {
+        $this->subscribeClient->subscribe(self::REDIS_KEY);
+        $this->subscribeClient->on('message', function ($channel, $payload) {
+            $this->onMessage($channel, $payload);
+        });
+    }
+
+    protected function onMessage($channel, $payload)
+    {
+        $payload = json_decode($payload);
+
+        if ($this->serverId === $payload->serverId) {
+            return false;
+        }
+
+        /* @var $channelManager ChannelManager */
+        $channelManager = app(ChannelManager::class);
+        $channelSearch        = $channelManager->find($payload->appId, $payload->channel);
+
+        if ($channelSearch === null) {
+            return false;
+        }
+
+        $channel->broadcast($payload);
```

Comment:
It's to support this feature: https://laravel.com/docs/5.7/broadcasting#broadcasting-events

If a socketId is provided in the payload, then `broadcastToEveryoneExcept` should be called. Essentially the point of this is for example in a chat client, a user sent out a message, there's no reason to send an event about that message back to the same user who just sent it, since the frontend already has that message in its state. It's just a minor optimization that makes it send N-1 messages instead of N (where N is number of connections in the channel)